src/sage/schemes/curves/plane_curve_arrangement.py

```diff
@@ -0,0 +1,1342 @@
+            return computed
+        if computed:
+            return computed
+        _ = self._fundamental_group(simplified=simplified)
+        if simplified:
+            return self._meridians_simpl
+        else:
+            return self._meridians_nonsimpl
+
+
+class PlaneCurveArrangements(UniqueRepresentation, Parent):
+    """
+    Plane curve arrangements.
+
+    INPUT:
+
+    - ``base_ring`` -- ring; the base ring
+
+    - ``names`` -- tuple of strings; the variable names
+
+    EXAMPLES::
+
+        sage: H.<x, y> = PlaneCurveArrangements(QQ)
+        sage: H(x, y^2, x-1, y-1)
+        Arrangement (x, y^2, x - 1, y - 1) in Affine Space
+        of dimension 2 over Rational Field
+    """
+    Element = PlaneCurveArrangementElement
+
+    @staticmethod
+    def __classcall__(cls, base, names=()):
+        """
+        Normalize names
+        TESTS::
+
+            sage: H.<x, y> = AffinePlaneCurveArrangements(QQ)
+            sage: K = AffinePlaneCurveArrangements(QQ, names=('x', 'y'))
+            sage: H is K
+            True
+            sage: TestSuite(K).run()
+        """
+        names = normalize_names(len(names), names)
+        return super().__classcall__(cls, base, names)
+
+    def __init__(self, base_ring, names=tuple()):
+        """
+        Initialize ``self``.
+
+        TESTS::
+
+            sage: H.<x, y> = AffinePlaneCurveArrangements(QQ)
+            sage: TestSuite(H).run()
+        """
+        if base_ring not in _Fields:
+            raise ValueError('base ring must be a field')
+        super().__init__(base_ring, names=names, category=Sets())
+        self._embedded = None
+        if len(names) == 2:
+            self._embedded = 'affine'
+        elif len(names) == 3:
+            self._embedded = 'projective'
+
+    def coordinate_ring(self):
+        """
+        Return the coordinate ring.
+
+        OUTPUT:
+
+        The coordinate ring of the curve arrangement.
+
+        EXAMPLES::
+
+            sage: L.<x, y> = AffinePlaneCurveArrangements(QQ)
+            sage: L.coordinate_ring()
+            Multivariate Polynomial Ring in x, y over Rational Field
+        """
+        return PolynomialRing(self.base_ring(), self.variable_names())
+
+    def change_ring(self, base_ring):
+        """
+        Return curve arrangements over a different base ring.
+
+        INPUT:
+
+        - ``base_ring`` -- a ring; the new base ring.
+
+        OUTPUT:
+
+        A new :class:`AffinePlaneCurveArrangements` instance over the new
+        base ring.
+
+        EXAMPLES::
+
+            sage: L.<x, y> = PlaneCurveArrangements(QQ)
+            sage: L.gen(0)
+            x
+            sage: L.change_ring(RR).base_ring()
+            Real Field with 53 bits of precision
+
+        TESTS::
+
+            sage: L.change_ring(QQ) is L
+            True
+        """
+        return PlaneCurveArrangements(base_ring, names=self.variable_names())
+
+    @cached_method
+    def ambient_space(self):
```

Comment:
I think this would be better as an `@abstract_method` and have each of the subclasses return the corresponding object. (It follows OOP principles better.)

Also, is there a benefit to caching this? The parents are unique IIRC. Is this called very frequently through some method?

Reply:
To be honest, I borrow this from the class or hyperplane arrangements: I just tookit out without putting `@abstract_method`, it gave plenty of errors

Reply:
The base class has the `@abstract_method` but the subclasses would override this (getting rid of the `if`) with `@cached_method`. Although I still don't quite understand the need for the caching.

Reply:
As a I said I copied from hyperplane arrangements. Maybe just taking out `@cached_method`'s is enough and maybe there is no need to add `@abstract_method`.

Reply:
It's not necessary, but the code is saying that you should have two methods: one for affine and one for projective (basically the `if` statement in there).

Reply:
No, I propose making it an `@abstract_method` (which will just have a docstring and a simple doctest) and then providing the concrete implementations (with proper doctests). (Although only having it in the subclasses, those other methods would work.)

Reply:
I am not sure if it is what you mean. I defined `ambient_space` as an `@abstract_method` for  `PlaneCurveArrangementElement`.  Then, I defined it for the subclasses. I got plenty of errors, e.g.,:
```
File "plane_curve_arrangement.py", line 8, in sage.schemes.curves.plane_curve_arrangement
Failed example:
    H.<x, y> = AffinePlaneCurveArrangements(QQ)
Exception raised:
    Traceback (most recent call last):
      File "/usr/local/sage-310/src/sage/doctest/forker.py", line 712, in _run
        self.compile_and_execute(example, compiler, test.globs)
      File "/usr/local/sage-310/src/sage/doctest/forker.py", line 1147, in compile_and_execute
        exec(compiled, globs)
      File "<doctest sage.schemes.curves.plane_curve_arrangement[0]>", line 1, in <module>
        H = AffinePlaneCurveArrangements(QQ, names=('x', 'y',)); (x, y,) = H._first_ngens(2)
                                                                           ^^^^^^^^^^^^^^^^^
      File "sage/structure/category_object.pyx", line 370, in sage.structure.category_object.CategoryObject._first_ngens
        names = self._defining_names()
      File "sage/misc/cachefunc.pyx", line 2329, in sage.misc.cachefunc.CachedMethodCallerNoArgs.__call__
        self.cache = f(self._instance)
      File "sage/structure/category_object.pyx", line 415, in sage.structure.category_object.CategoryObject._defining_names
        return self.gens()
      File "sage/misc/cachefunc.pyx", line 2329, in sage.misc.cachefunc.CachedMethodCallerNoArgs.__call__
        self.cache = f(self._instance)
      File "/usr/local/sage-310/src/sage/schemes/curves/plane_curve_arrangement.py", line 1256, in gens
        return self.ambient_space().gens()
               ^^^^^^^^^^^^^^^^^^
      File "/usr/local/sage-310/src/sage/misc/abstract_method.py", line 220, in __get__
        raise NotImplementedError(repr(self))
    NotImplementedError: <abstract method ambient_space at 0x7f05de12f920>
```
I do not know how to continue

Reply:
I saw that I misplaced the methods, the abstract method should be in `PlaneCurveArrangement` if I am not wong (and the specific ones in `AffinePlaneCurveArrangement`  and `ProjectivePlaneCurveArrangement` ) but the result is the same

Reply:
Then either you are doing something wrong or its exposing a bug. Perhaps try it without the `@cached_method` (although I would be *very* surprised if that worked, but that is a potential bug)? Also, please push what you have, otherwise I am stuck guessing at what you're trying.

Reply:
Sorry, I erased the last changes. Anyway, `ambient_space` is not a `@cached_method` any more at this moment. There are three other `@cached_method`'s but there is no deep reason, they could be wiped out.
I can also try to use `@abstract_method` again; I wonder if the errors come from the appearance of `ambient_space` in several methods (`__init__` and `_element_constructor_` at least)

Reply:
Definitely not `_element_constructor_`, and I am almost certain for `__init__` with how Python works. Although I would need to see the code you are trying, and possibly play around with it myself, to see what is going on.

You are basically saying the standard OOP doesn't work otherwise, which Python very much promises and I don't believe.

Reply:
Push done

Reply:
Thanks. I will take a look tonight after dinner.

Reply:
I see. Your tests using `PlaneCurveArrangement` are now bad. By adding `@abstract_method`, you make the class into an *abstract* class, which means it doesn't expect to be able to be constructed on its own. Instead you must create/use a specific (concrete) subclass. Just simply change the `PlaneCurveArrangement` into one of the affine or projective arrangements in the doctests. The code works for me otherwise it seems.

Reply:
I change the doctests and I made also `change_ring` for `PlaneCurveArrangement` an abstract method. Not sure if the doctests for both abstract methods are the needed one.

Reply:
Thank you.

Annoyingly yes...but at least it is simple to construct them.

Reply:
In the PR page it is said that there are requested changes but I find no one which was not treated.